pkg/querier/batch/stream_test.go

```diff
@@ -19,128 +19,161 @@ import (
 	"github.com/grafana/mimir/pkg/util/test"
 )
 
-func TestStream(t *testing.T) {
+func mkFloatBatch(from int64) chunk.Batch {
+	return mkGenericFloatBatch(from, chunk.BatchSize)
+}
+
+func mkHistogramBatch(from int64) chunk.Batch {
+	return mkGenericHistogramBatch(from, chunk.BatchSize)
+}
+
+func mkGenericFloatBatch(from int64, size int) chunk.Batch {
+	batch := chunk.Batch{ValueType: chunkenc.ValFloat}
+	for i := 0; i < size; i++ {
+		batch.Timestamps[i] = from + int64(i)
+		batch.Values[i] = float64(from + int64(i))
+	}
+	batch.Length = size
+	return batch
+}
+
+func mkGenericHistogramBatch(from int64, size int) chunk.Batch {
+	batch := chunk.Batch{ValueType: chunkenc.ValHistogram}
+	for i := 0; i < size; i++ {
+		batch.Timestamps[i] = from + int64(i)
+		batch.PointerValues[i] = unsafe.Pointer(test.GenerateTestHistogram(int(from) + i))
+	}
+	batch.Length = size
+	return batch
+}
+
+func requireBatchEqual(t *testing.T, b, o chunk.Batch) {
+	require.Equal(t, b.ValueType, o.ValueType)
+	require.Equal(t, b.Length, o.Length)
+	for i := 0; i < b.Length; i++ {
+		switch b.ValueType {
+		case chunkenc.ValFloat:
+			require.Equal(t, b.Values[i], o.Values[i], fmt.Sprintf("at idx %v", i))
+		case chunkenc.ValHistogram:
+			bh := (*histogram.Histogram)(b.PointerValues[i])
+			oh := (*histogram.Histogram)(o.PointerValues[i])
+			require.Equal(t, *bh, *oh, fmt.Sprintf("at idx %v", i))
+		case chunkenc.ValFloatHistogram:
+			bh := (*histogram.FloatHistogram)(b.PointerValues[i])
+			oh := (*histogram.FloatHistogram)(o.PointerValues[i])
+			require.Equal(t, *bh, *oh, fmt.Sprintf("at idx %v", i))
+		}
+	}
+}
+
+func TestBatchStream_Empty(t *testing.T) {
+	s := newBatchStream(1, nil, nil)
+	b1 := mkHistogramBatch(0)
+	b2 := mkHistogramBatch(chunk.BatchSize)
+	s.batches = []chunk.Batch{b1, b2}
+
+	require.Len(t, s.batches, 2)
+	require.Equal(t, b1, s.batches[0])
+	require.Equal(t, b2, s.batches[1])
+
+	s.empty()
+	require.Len(t, s.batches, 0)
+}
+
+func TestBatchStream_RemoveFirst(t *testing.T) {
+	s := newBatchStream(1, nil, nil)
+	b1 := mkHistogramBatch(0)
+	b2 := mkHistogramBatch(chunk.BatchSize)
+	s.batches = []chunk.Batch{b1, b2}
+
+	require.Len(t, s.batches, 2)
+	require.Equal(t, b1, s.batches[0])
+	require.Equal(t, b2, s.batches[1])
+
+	s.removeFirst()
+	require.Len(t, s.batches, 1)
+	require.Equal(t, b2, s.batches[0])
+}
+
+func TestBatchStream_Merge(t *testing.T) {
 	for i, tc := range []struct {
-		testcase       string
-		input1, input2 []chunk.Batch
-		output         batchStream
+		testcase string
+		batches  []chunk.Batch
+		batch    chunk.Batch
```

Comment:
I think we should keep the existing test cases where we have multiple inputs on the "right" side, the test just has to iterate through them and merge one by one

Reply:
I have applied your suggestion.